### PR TITLE
Added a fixer for casing of variables and properties

### PR DESCRIPTION
**Pull request by kierate**

### CODE REVIEW

File: Symfony/CS/Tests/Fixer/Contrib/VariablePropertyCaseFixerTest.php [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            ),
+            array(
+                '<?php public function do_something() {}',
+            ),
+            array(
+                '<?php $this->some_function();',
+            ),
+            array(
+                '<?php private $someVar = "foo";',
+                '<?php private $some_var = "foo";',
```
Comment (stof) [inline Symfony/CS/Tests/Fixer/Contrib/VariablePropertyCaseFixerTest.php]: this is not valid PHP code

File: Symfony/CS/Tests/Fixer/Contrib/VariablePropertyCaseFixerTest.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    public function provideFixCases()
+    {
+        return array(
+            array(
+                '<?php class foo_bar {}',
+            ),
+            array(
+                '<?php public function do_something() {}',
```
Comment (stof) [inline Symfony/CS/Tests/Fixer/Contrib/VariablePropertyCaseFixerTest.php]: this is not valid PHP code

**Comment by stof** [comment]
This fixer must be documented as being unsafe (changing public and protected properties can break code for other places)

**Comment by GrahamCampbell** [comment]
> This fixer must be documented as being unsafe (changing public and protected properties can break code for other places)

TBH, this fixer looks like a bad idea for that reason alone. I think other tools that can highlight, but not "fix" variables might be more appropriate.

**Comment by kierate** [comment]
@stof I was going for minimalism with the original test examples:) but have extended them a bit now. The fixer is unsafe, so I have updated the description with a warning.

@GrahamCampbell I do understand that most people would first use something else to highlight the problems. The fixer though is intended to save time on "manual labour".

File: Symfony/CS/Fixer/Contrib/VariablePropertyCaseFixer.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+     */
+    private function fixUnderscoreToCamelCase(Token $token)
+    {
+        if (strstr($token->getContent(), '_') !== false) {
+            $token->override(
+                Utils::underscoreToCamelCase($token->getContent())
+            );
+        }
+
+        return;
```
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/VariablePropertyCaseFixer.php]: what for ?

Comment (GrahamCampbell) [inline Symfony/CS/Fixer/Contrib/VariablePropertyCaseFixer.php]: I assume it used to be "return null;" before php-cs-fixer was run on itself, lol.

**Comment by keradus** [comment]
For now we have few dengerous fixers like `strict` or `ereg_to_preg`. All of them help to deal with deprecated syntax or loosely typing. All of them will change the code in context of file. Any changes made by them will affects only that one file.

Here you are changing one file, and affects all codebase. When changing your own property you are breaking your API. But what is more - you also changing usage of 3rd party code (eg in case of inheritance), and this changes simply MUST NOT be done.

:-1:

**Comment by kierate** [comment]
This fixer is in contrib so you have to enable it, right?
If that is the case then surely whoever decides to use this as a tool to simplify their task (of switching from underscores to camel case) is well aware of the repercussions of the task they are undertaking? I don't understand why not allowing this in the tool will help with the problem you are describing (especially if you explicitly have to say "yes, fix the variables").
Just something to consider... I found this useful, thought someone else might as well, but no biggie if you want to reject this.

**Comment by keradus** [comment]
> This fixer is in contrib so you have to enable it, right?

Yes, but still - this will probably destroy your code - this mean that you will need to do sth more to made it work again.


> I don't understand why not allowing this in the tool will help with the problem you are describing

Without this fixer there would not be this problem at all. 
After adding this fixer... first - a lot of dissatisfaction will occurs about need of manually fix after auto fix (In one file I will have `$this->my_var` fixed and in other `$instance->my_var` not.), second - sb will need to maintaint this fixer.

**Comment by kierate** [comment]
@keradus I am not arguing the concept of *potenially* having to fix outside code. That totally might happen.
The thing I am trying to put across that as a tool this fixer does a particular job automatically, which saves manual work if you have to do convert code from underscores to camel case.

I don't really see a massive conceptual difference between running this unsafe fixer and running the `strict` fixer on something like this:
```php
class Foo
{
    public function isEqual($object)
    {
        return $this->bar == $object;
    }
}
```
and then having some outside code that uses it e.g.:
```php
$bar = ...; //some object
$foo = new Foo();
var_dump($foo->isEqual($bar));
```
In this case the result of isEqual can (and in a lot of cases will) be different i.e. the `strict` fixer is breaking how the outside code functions.

The same applies to my fixer albeit I would say that in cases where it does break anything it at least does it in an obvious way, rather then "silently" changing behaviour.

With regards to your comment about dissatisfaction - all I can say is that I hope that no sane developer would explicitly run this on their own codebase and then complain to you that something is broken...

**Comment by keradus** [comment]
Working preg_to_ereg etc works on one file and it works great. All current fixers help to deal with deprecated syntax or loosely typing, as I said before.
Here, you cant work on single file without full context of codebase.

**Comment by kierate** [comment]
@keradus yeah I get that `preg_to_ereg` won't change any outside code, but I was talking about the `strict` fixer, which can affect code outside of your codebase. IMO this is totally fine, any developer who decides to make that change should be aware of that risk.

From the second sentence of the readme for this project: "If you are already using PHP_CodeSniffer to identify coding standards problems in your code, **you know that fixing them by hand is tedious, especially on large projects. This tool does the job for you.**". Not adding this fixer just means that no one will be able to use your tool to help fix variable casing - they will instead have the tedious task of fixing them by hand...

**Comment by keradus** [comment]
IMO the level of risk this will introduce is much greater than `strict` or `preg_to_ereg`.
I am against this fixer and already said why, no need to repeat. I bet @fabpot will be against it too.

Any other thoughts ?

**Comment by sstok** [comment]
:-1: For the same reasons keradus mentioned. Even when they are private it may cause issues with say Doctrine Entity mapping. The risk of breakage is to high.

**Comment by stof** [comment]
:-1: as well (same reasons than above)

**Comment by keradus** [comment]
Closing due to discussion.

**Pull request closed without merging**